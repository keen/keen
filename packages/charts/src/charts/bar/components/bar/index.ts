import Bar from './bar.component';

export default Bar;
