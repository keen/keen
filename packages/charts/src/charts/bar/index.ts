import BarChart, { Props as BarChartSettings } from './bar-chart.component';

export { BarChart };
export type { BarChartSettings };
