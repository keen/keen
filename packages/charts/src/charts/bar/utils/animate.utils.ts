import { Layout } from '@keen.io/ui-core';

import { GroupMode } from '../../../types';

type BarMotionProps = {
  x: number;
  y: number;
  height: number;
  width: number;
  groupMode: GroupMode;
  layout: Layout;
};

/**
 * Prepare animation for bar
 *
 * @param groupMode - type of groupMode
 * @param layout - layout of the chart
 * @param height - height of the bar
 * @param width - width of the bar
 * @param x - x position of the bar
 * @param y - y position of the bar
 * @return animation object with data for framer
 *
 */

export const createBarMotion = ({
  groupMode,
  layout,
  height,
  width,
  x,
  y,
}: BarMotionProps) => {
  const motion = {
    exit: {
      opacity: 0,
    },
    animate: {
      x,
      y,
      height,
      width,
      opacity: 1,
      transition: {
        duration: 0.4,
      },
    },
    active: {
      opacity: 1,
    },
    inactive: {
      opacity: 0.2,
    },
  };

  if (groupMode === 'grouped') {
    return {
      ...motion,
      initial: {
        opacity: 0,
        ...(layout === 'vertical'
          ? {
              x,
              y: height + y,
              height: 0,
            }
          : {
              y,
              x,
              width: 0,
            }),
      },
    };
  } else {
    return {
      ...motion,
      initial: {
        opacity: 0,
        x,
        y,
      },
    };
  }
};
