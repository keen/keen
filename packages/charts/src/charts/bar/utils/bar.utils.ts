import { transparentize } from 'polished';
import { ScaleBand, ScaleLinear, ScaleTime } from 'd3-scale';
import { Layout, SortMode } from '@keen.io/ui-core';
import { getPaletteColor } from '@keen.io/charts-utils';

import { GroupMode, StackMode, DataSelector } from '../../../types';
import { Bar } from '../types';

/**
 * Prepare a color based on groupMode and stackMode
 *
 * @param activeBar - selected bar with mouse hover
 * @param barKey - key of the bar
 * @param barSelector - selector of the bar
 * @param stackMode - stackMode option
 * @param groupMode - groupMode option
 * @param color - color of the bar
 * @return return a color
 *
 */

export const getBarColor = ({
  activeBar,
  barKey,
  barSelector,
  stackMode,
  groupMode,
  color,
}: {
  groupMode: GroupMode;
  stackMode: StackMode;
  color: string;
  barKey: string;
  barSelector: DataSelector;
  activeBar: {
    selector: DataSelector;
    key: string;
  };
}) => {
  if (groupMode === 'stacked' && stackMode === 'normal') {
    const [, activeProperty] = activeBar.selector;
    const [, barProperty] = barSelector;

    if (activeProperty === barProperty) {
      return color;
    }

    if (activeProperty) {
      return transparentize(0.4, color);
    }

    return color;
  }

  if (groupMode === 'stacked' && stackMode === 'percent') {
    const [activeIndex] = activeBar.selector;
    const [barIndex] = barSelector;

    if (activeIndex === barIndex) {
      return transparentize(0.05, color);
    }

    return color;
  }

  return activeBar.key === barKey ? transparentize(0.05, color) : color;
};

/**
 * Calculate bars data for chart
 *
 * @param data - data series
 * @param groupScale - band scale
 * @param keys - keys used for calculation
 * @param labelSelector - selected label from data
 * @param disabledKeys - keys disabled for calculation/display
 * @param range - array of booleans created based on yScale/xScale domain length
 * @param xScale - time/linear scale
 * @param yScale - time/linear scale
 * @param layout - plot layout
 * @param colors - array of colors
 * @param barsOrder - bars order
 * @return bars for chart
 *
 */

export const calculateGroupedBars = (
  data: Record<string, any>[],
  groupScale: ScaleBand<string>,
  keys: string[],
  labelSelector: string,
  disabledKeys: string[],
  range: boolean[],
  xScale:
    | ScaleBand<string | number>
    | ScaleLinear<number, number>
    | ScaleTime<number, number>,
  yScale: ScaleBand<string | number> | ScaleLinear<number, number>,
  layout: Layout,
  colors: string[],
  barsOrder?: SortMode
) => {
  const barSize = groupScale.bandwidth();
  const keysOrder: Record<number, any> = {};

  if (barsOrder) {
    data.forEach((_item: Record<string, any>, idx: number) => {
      groupScale
        .domain()
        .sort((a, b) => {
          if (barsOrder === 'descending') {
            return data[idx]?.[b] - data[idx]?.[a];
          } else {
            return data[idx]?.[a] - data[idx]?.[b];
          }
        })
        .filter((keyName: string) => data[idx][keyName])
        .forEach((keyName: string, index: number) => {
          if (!keysOrder[idx]) keysOrder[idx] = {};
          keysOrder[idx][keyName] = index;
        });
    });
  }
  const bars = [] as Bar[];
  range.forEach((_d, index: number) => {
    let counter = 0;
    const barsToRenderInGroup = Object.entries(data[index]).filter(
      (el) => el[1] !== 0 && el[0] !== labelSelector
    ).length;

    const newOrder = (keys.length - barsToRenderInGroup) / 2;

    keys.forEach((keyName: string, idx: number) => {
      const isDisabled =
        keyName === labelSelector || disabledKeys.includes(keyName);

      const value = data[index]?.[keyName];
      if (value && !isDisabled) {
        const orderPosition = barsOrder
          ? keysOrder[index][keyName] + newOrder
          : counter + newOrder;

        const bar =
          layout === 'vertical'
            ? {
                x: xScale(data[index][labelSelector]) + barSize * orderPosition,
                y: value > 0 ? yScale(value) : yScale(0),
                width: barSize,
                height: Math.abs(yScale(value) - yScale(0)),
              }
            : {
                x: value > 0 ? Math.abs(xScale(0)) : xScale(value),
                y: yScale(data[index][labelSelector]) + barSize * orderPosition,
                width: Math.abs(xScale(value) - xScale(0)),
                height: barSize,
              };

        bars.push({
          key: `${index}.${keyName}`,
          selector: [index, keyName],
          color: getPaletteColor(idx, colors),
          value,
          ...bar,
        });
      }
      if (!barsOrder && !disabledKeys.includes(keyName) && value) counter++;
    });
  });

  return bars;
};
