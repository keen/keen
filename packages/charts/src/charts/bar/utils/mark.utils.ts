import { Layout } from '@keen.io/ui-core';

type MarkVariants = 'normal' | 'small';

export const MARK_VARIANTS: Record<MarkVariants, Record<string, any>> = {
  normal: {
    outerRadius: 10,
    innerRadius: 4,
    strokeWidth: 4,
  },
  small: {
    outerRadius: 7,
    innerRadius: 3,
    strokeWidth: 3,
  },
};

const MARK_TRESHOLD = 35;

/**
 * Calculate mark position based on bar data
 *
 * @param layout - layout of the chart
 * @param x - x position of the bar
 * @param y - y position of the bar
 * @param width - width of the bar
 * @param height - height of the bar
 * @param positiveValue - if bar have positive value to display mark correctly
 * @return x and y for mark position
 *
 */

export const calculateMarkPosition = ({
  layout,
  x,
  y,
  width,
  height,
  positiveValue,
}: {
  layout: Layout;
  x: number;
  y: number;
  width: number;
  height: number;
  positiveValue: boolean;
}) => {
  if (layout === 'vertical') {
    return {
      x: x + width * 0.5,
      y: positiveValue ? y : y + height,
    };
  }

  return {
    x: positiveValue ? x + width : x,
    y: y + height * 0.5,
  };
};

/**
 * Set mark size based on plot layout and width/height of the bar
 *
 * @param layout - layout of the chart
 * @param width - width of the bar
 * @param height - height of the bar
 * @return x and y for mark position
 *
 */

export const setMarkSize = ({
  layout,
  height,
  width,
}: {
  layout: Layout;
  width: number;
  height: number;
}) => {
  if (layout === 'vertical') {
    return width < MARK_TRESHOLD
      ? MARK_VARIANTS['small']
      : MARK_VARIANTS['normal'];
  }
  return height < MARK_TRESHOLD
    ? MARK_VARIANTS['small']
    : MARK_VARIANTS['normal'];
};
