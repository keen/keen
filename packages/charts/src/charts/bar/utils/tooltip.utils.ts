import { getFromPath } from '@keen.io/charts-utils';

import {
  DataSelector,
  GroupMode,
  StackMode,
  TooltipFormatter,
} from '../../../types';

export const getLabel = ({
  isPercentage,
  selector,
  data,
  percentageData,
  formatValue,
}: {
  selector: DataSelector;
  isPercentage: boolean;
  data: Record<string, any>[];
  percentageData: Record<string, any>[];
  formatValue?: TooltipFormatter;
}) => {
  const value = formatValue
    ? formatValue(getFromPath(data, selector))
    : getFromPath(data, selector);
  if (isPercentage) {
    return {
      value,
      change: `(${getFromPath(percentageData, selector).toFixed(2)}%)`,
    };
    // return `${value} (${getFromPath(percentageData, selector).toFixed(2)}%)`;
  }

  return { value };
};

export const getSelectors = ({
  groupMode,
  stackMode,
  keys,
  disabledKeys,
  colors,
  selector,
}: {
  groupMode: GroupMode;
  stackMode: StackMode;
  disabledKeys: string[];
  keys: string[];
  colors: string[];
  selector: { selector: DataSelector; color: string };
}): { selector: DataSelector; color: string }[] => {
  if (groupMode === 'stacked' && stackMode === 'percent') {
    // if (groupMode === 'stacked') {
    const { selector: dataSelector } = selector;
    const [index] = dataSelector;
    const selectors: { selector: DataSelector; color: string }[] = [];

    keys.forEach((key: string, idx: number) => {
      if (!disabledKeys.includes(key)) {
        selectors.push({
          selector: [index, key],
          color: colors[idx],
        });
      }
    });

    return selectors;
  }

  return [selector];
};
