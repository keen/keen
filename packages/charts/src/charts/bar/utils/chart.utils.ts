import { scaleBand, scaleLinear } from 'd3-scale';
import { stack, stackOffsetDiverging } from 'd3-shape';

import { Layout, SortMode } from '@keen.io/ui-core';
import {
  calculateRange,
  calculateStackedRange,
  calculateScaleDomain,
  getKeysDifference,
  transformToPercent,
  normalizeDate,
  ScaleSettings,
  getOffsetRangeColor,
} from '@keen.io/charts-utils';

import { calculateGroupedBars } from './bar.utils';

import { Bar } from '../types';
import { Dimension, Margins, GroupMode, StackMode } from '../../../types';

type Options = {
  data: Record<string, any>[];
  keys: string[];
  disabledKeys: string[];
  labelSelector: string;
  dimension: Dimension;
  margins: Margins;
  layout: Layout;
  barPadding: number;
  groupMode: GroupMode;
  stackMode: StackMode;
  minValue?: number | 'auto';
  maxValue?: number | 'auto';
  colors: string[];
  xScaleSettings: ScaleSettings;
  yScaleSettings: ScaleSettings;
  xAxisTitle?: string;
  yAxisTitle?: string;
  barsOrder?: SortMode;
  dataSeriesOffset?: [number, number];
};

/**
 * Generate all data needed to set up grouped horizontal bar chart
 *
 * @param data - data series
 * @param keys - keys used in calculations
 * @param disabledKeys - keys disabled for calculation/display
 * @param dimension - predefine dimensions of the chart
 * @param margins - predefine margins of the chart
 * @param minValue - predefine minimum value
 * @param maxValue - predefine maximum value
 * @param barPadding - padding between bars
 * @param barsOrder - order of the bars
 * @param labelSelector - selected label from data
 * @param colors - palette of colors for chart
 * @param xScaleSettings - time scale settings
 * @param layout - layout of the plot
 * @return bars, xScale, yScale, localizedData
 *
 */

export const generateHorizontalGroupedBars = ({
  data,
  keys,
  disabledKeys,
  dimension,
  margins,
  minValue,
  maxValue,
  barPadding,
  barsOrder,
  labelSelector,
  colors,
  xScaleSettings,
  layout,
  dataSeriesOffset,
}: Options) => {
  const filteredKeys = getKeysDifference(keys, disabledKeys);
  const { precision } = xScaleSettings;

  const dateNormalizer = precision
    ? (date: string) => normalizeDate(date, precision)
    : null;

  const localizedData = dateNormalizer
    ? data.map((item) => ({
        ...item,
        [labelSelector]: dateNormalizer(item[labelSelector]),
      }))
    : data;

  const { minimum, maximum } = calculateRange(
    localizedData,
    minValue,
    maxValue,
    keys
  );

  const xScale = scaleLinear()
    .range([margins.left, dimension.width - margins.right])
    .domain([minimum, maximum]);

  calculateScaleDomain(xScale, minimum, maximum);

  const yScale = scaleBand()
    .range([dimension.height - margins.bottom, margins.top])
    .domain(localizedData.map((item: any) => item[labelSelector]).reverse())
    .padding(barPadding);

  const yGroupScale = scaleBand()
    .range([0, yScale.bandwidth()])
    .domain(filteredKeys);

  const range = new Array(yScale.domain().length).fill(true);

  const bars = calculateGroupedBars(
    localizedData,
    yGroupScale,
    keys,
    labelSelector,
    disabledKeys,
    range,
    xScale,
    yScale,
    layout,
    colors,
    barsOrder,
    dataSeriesOffset
  );

  return {
    bars,
    xScale,
    yScale,
    localizedData,
  };
};

/**
 * Generate all data needed to set up grouped vertical bar chart
 *
 * @param data - data series
 * @param dimension - predefine dimensions of the chart
 * @param margins - predefine margins of the chart
 * @param minValue - predefine minimum value
 * @param maxValue - predefine maximum value
 * @param barPadding - padding between bars
 * @param barsOrder - order of the bars
 * @param keys - keys used in calculations
 * @param disabledKeys - keys disabled for calculation/display
 * @param colors - palette of colors for chart
 * @param labelSelector - selected label from data
 * @param xScaleSettings - time scale settings
 * @param layout - layout of the plot
 * @return bars, xScale, yScale, localizedData
 *
 */

export const generateVerticalGroupedBars = ({
  data,
  dimension,
  margins,
  minValue,
  maxValue,
  barPadding,
  barsOrder,
  keys,
  disabledKeys,
  colors,
  labelSelector,
  xScaleSettings,
  layout,
  dataSeriesOffset,
}: Options) => {
  const filteredKeys = getKeysDifference(keys, disabledKeys);
  const { precision } = xScaleSettings;

  const dateNormalizer = precision
    ? (date: string) => normalizeDate(date, precision)
    : null;

  const localizedData = dateNormalizer
    ? data.map((item) => ({
        ...item,
        [labelSelector]: dateNormalizer(item[labelSelector]),
      }))
    : data;

  const { minimum, maximum } = calculateRange(
    localizedData,
    minValue,
    maxValue,
    keys
  );

  const xScale = scaleBand()
    .range([margins.left, dimension.width - margins.right])
    .domain(localizedData.map((item: any) => item[labelSelector]))
    .padding(barPadding);

  const yScale = scaleLinear()
    .range([dimension.height - margins.bottom, margins.top])
    .domain([minimum, maximum]);

  calculateScaleDomain(yScale, minimum, maximum);

  const xGroupScale = scaleBand()
    .range([0, xScale.bandwidth()])
    .domain(filteredKeys);

  const range = new Array(xScale.domain().length).fill(true);

  const bars = calculateGroupedBars(
    localizedData,
    xGroupScale,
    keys,
    labelSelector,
    disabledKeys,
    range,
    xScale,
    yScale,
    layout,
    colors,
    barsOrder,
    dataSeriesOffset
  );

  return {
    bars,
    xScale,
    yScale,
    localizedData,
  };
};

/**
 * Generate all data needed to set up stacked horizontal bar chart
 *
 * @param data - data series
 * @param dimension - predefine dimensions of the chart
 * @param margins - predefine margins of the chart
 * @param minValue - predefine minimum value
 * @param maxValue - predefine maximum value
 * @param barPadding - padding between bars
 * @param keys - keys used in calculations
 * @param disabledKeys - keys disabled for calculation/display
 * @param colors - palette of colors for chart
 * @param stackMode - stackMode option
 * @param labelSelector - selected label from data
 * @param xScaleSettings - time scale settings
 * @return bars, xScale, yScale, localizedData
 *
 */

export const generateHorizontalStackedBars = ({
  data,
  dimension,
  margins,
  minValue,
  maxValue,
  barPadding,
  keys,
  disabledKeys,
  colors,
  stackMode,
  labelSelector,
  xScaleSettings,
  dataSeriesOffset,
}: Options) => {
  const bars = [] as Bar[];
  const filteredKeys = getKeysDifference(keys, disabledKeys);
  const { precision } = xScaleSettings;

  const dateNormalizer = precision
    ? (date: string) => normalizeDate(date, precision)
    : null;

  const localizedData = dateNormalizer
    ? data.map((item) => ({
        ...item,
        [labelSelector]: dateNormalizer(item[labelSelector]),
      }))
    : data;

  const normalizedData =
    stackMode === 'normal'
      ? localizedData
      : transformToPercent(localizedData, filteredKeys);

  const stackedData = stack().keys(filteredKeys).offset(stackOffsetDiverging)(
    normalizedData
  );

  let { minimum, maximum } = calculateStackedRange(
    normalizedData,
    minValue,
    maxValue,
    filteredKeys
  );

  const percentMin = minimum < -100 ? -100 : minimum;
  const percentMax = maximum > 100 ? 100 : maximum;

  minimum = stackMode === 'percent' ? percentMin : minimum;
  maximum = stackMode === 'percent' ? percentMax : maximum;

  const xScale = scaleLinear()
    .range([margins.left, dimension.width - margins.right])
    .domain([minimum, maximum]);

  const yScale = scaleBand()
    .range([dimension.height - margins.bottom, margins.top])
    .domain(normalizedData.map((item: any) => item[labelSelector]).reverse())
    .padding(barPadding);

  calculateScaleDomain(xScale, minimum, maximum);

  const barHeight = yScale.bandwidth();
  const range = new Array(yScale.domain().length).fill(true);

  stackedData.forEach((item: any) => {
    const keyName = item.key;

    range.forEach((_d, index: number) => {
      const [rangeMin, rangeMax] = item[index];
      const width = xScale(rangeMax) - xScale(rangeMin) || 0;

      const bar = {
        key: `${index}.${keyName}`,
        selector: [index, keyName],
        x: xScale(rangeMin),
        y: yScale(normalizedData[index][labelSelector]),
        width,
        height: barHeight,
        color: getOffsetRangeColor(
          keys.indexOf(keyName),
          colors,
          dataSeriesOffset
        ),
        colorOutOfRange: dataSeriesOffset
          ? index < dataSeriesOffset[0] || index >= dataSeriesOffset[1]
          : !colors[index],
        value: data[index][keyName],
      };

      bars.push(bar);
    });
  });

  return {
    bars,
    xScale,
    yScale,
    localizedData,
  };
};

/**
 * Generate all data needed to set up stacked vertical bar chart
 *
 * @param data - data series
 * @param dimension - predefine dimensions of the chart
 * @param margins - predefine margins of the chart
 * @param minValue - predefine minimum value
 * @param maxValue - predefine maximum value
 * @param barPadding - padding between bars
 * @param keys - keys used in calculations
 * @param disabledKeys - keys disabled for calculation/display
 * @param colors - palette of colors for chart
 * @param stackMode - stackMode option
 * @param labelSelector - selected label from data
 * @param xScaleSettings - time scale settings
 * @return bars, xScale, yScale, localizedData
 *
 */

export const generateVerticalStackedBars = ({
  data,
  dimension,
  margins,
  minValue,
  maxValue,
  barPadding,
  keys,
  disabledKeys,
  colors,
  stackMode,
  labelSelector,
  xScaleSettings,
  dataSeriesOffset,
}: Options) => {
  const bars = [] as Bar[];
  const filteredKeys = getKeysDifference(keys, disabledKeys);
  const { precision } = xScaleSettings;

  const dateNormalizer = precision
    ? (date: string) => normalizeDate(date, precision)
    : null;

  const localizedData = dateNormalizer
    ? data.map((item) => ({
        ...item,
        [labelSelector]: dateNormalizer(item[labelSelector]),
      }))
    : data;

  const normalizedData =
    stackMode === 'normal'
      ? localizedData
      : transformToPercent(localizedData, filteredKeys);

  const stackedData = stack().keys(filteredKeys).offset(stackOffsetDiverging)(
    normalizedData
  );

  let { minimum, maximum } = calculateStackedRange(
    normalizedData,
    minValue,
    maxValue,
    filteredKeys
  );

  const percentMin = minimum < -100 ? -100 : minimum;
  const percentMax = maximum > 100 ? 100 : maximum;

  minimum = stackMode === 'percent' ? percentMin : minimum;
  maximum = stackMode === 'percent' ? percentMax : maximum;

  const xScale = scaleBand()
    .range([margins.left, dimension.width - margins.right])
    .domain(normalizedData.map((item: any) => item[labelSelector]))
    .padding(barPadding);

  const yScale = scaleLinear()
    .range([dimension.height - margins.bottom, margins.top])
    .domain([minimum, maximum]);

  calculateScaleDomain(yScale, minimum, maximum);

  const barWidth = xScale.bandwidth();
  const range = new Array(xScale.domain().length).fill(true);

  stackedData.forEach((item: any) => {
    const keyName = item.key;

    range.forEach((_d, index: number) => {
      const [rangeMin, rangeMax] = item[index];
      const height = yScale(rangeMin) - yScale(rangeMax) || 0;

      const bar = {
        key: `${index}.${keyName}`,
        selector: [index, keyName],
        x: xScale(normalizedData[index][labelSelector]),
        y: yScale(rangeMax),
        width: barWidth,
        height,
        color: getOffsetRangeColor(
          keys.indexOf(keyName),
          colors,
          dataSeriesOffset
        ),
        colorOutOfRange: dataSeriesOffset
          ? index < dataSeriesOffset[0] || index >= dataSeriesOffset[1]
          : !colors[index],
        value: data[index][keyName],
      };

      bars.push(bar);
    });
  });

  return {
    bars,
    xScale,
    yScale,
    localizedData,
  };
};

const BAR_RENDER_MAP: Record<string, any> = {
  grouped: {
    vertical: generateVerticalGroupedBars,
    horizontal: generateHorizontalGroupedBars,
  },
  stacked: {
    vertical: generateVerticalStackedBars,
    horizontal: generateHorizontalStackedBars,
  },
};

/**
 * Prepare options and check whitch function will be used based on layout and groupMode
 *
 * @param options - all options used in function generateVerticalGroupedBars or/and
 * generateHorizontalGroupedBars or/and generateVerticalStackedBars or/and
 * generateHorizontalStackedBars
 * @return function to use for generating bar chart data
 *
 */

export const generateBars = (options: Options) => {
  const {
    layout,
    groupMode,
    xScaleSettings,
    yScaleSettings,
    xAxisTitle,
    yAxisTitle,
  } = options;

  const settings =
    layout === 'vertical'
      ? {
          xScaleSettings,
          yScaleSettings,
          xAxisTitle,
          yAxisTitle,
        }
      : {
          xScaleSettings: yScaleSettings,
          yScaleSettings: xScaleSettings,
          xAxisTitle: yAxisTitle,
          yAxisTitle: xAxisTitle,
        };

  return {
    settings,
    ...BAR_RENDER_MAP[groupMode][layout].call(null, options),
  };
};
