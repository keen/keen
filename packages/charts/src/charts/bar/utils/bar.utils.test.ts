import { scaleBand, scaleLinear } from 'd3-scale';
import { getBarColor, getColor, calculateGroupedBars } from './bar.utils';

import { chartData } from '../bar-chart.fixtures';

const colors = ['red', 'green', 'blue'];

describe('getBarColor()', () => {
  test('should get correct bar color', () => {
    const parameters = {
      activeBar: {
        selector: [1, 'users'],
        key: '1.users',
      },
      barKey: '1.users',
      barSelector: [1, 'users'],
      stackMode: 'normal',
      groupMode: 'grouped',
      color: 'red',
    };
    const result = getBarColor(parameters);
    expect(result).toBe('rgba(255,0,0,0.95)');
  });
});

describe('getColor()', () => {
  test('should return green color', () => {
    const result = getColor(1, colors);

    expect(result).toBe('green');
  });

  test('should return black color', () => {
    const result = getColor(3, colors);

    expect(result).toBe('#1D2729');
  });
});

describe('calculateGroupedBars()', () => {
  const keys = ['users', 'licenses', 'shops'];
  const labelSelector = 'name';
  const range = [true, true, true, true];

  test('should calculate grouped bars correctly for vertical layout', () => {
    const xScale = scaleBand()
      .range([0, 100])
      .domain(chartData.map((item: any) => item[labelSelector]))
      .padding(0);
    const yScale = scaleLinear().range([0, 100]).domain([-30, 82]);
    const groupScale = scaleBand().range([0, xScale.bandwidth()]).domain(keys);

    const result = calculateGroupedBars(
      chartData,
      groupScale,
      keys,
      labelSelector,
      [],
      range,
      xScale,
      yScale,
      'vertical',
      colors
    );

    expect(result).toMatchSnapshot();
  });

  test('should calculate grouped bars correctly for horizontal layout', () => {
    const xScale = scaleLinear().range([0, 100]).domain([-30, 82]);
    const yScale = scaleBand()
      .range([0, 100])
      .domain(chartData.map((item: any) => item[labelSelector]).reverse())
      .padding(0);
    const groupScale = scaleBand().range([0, yScale.bandwidth()]).domain(keys);

    const result = calculateGroupedBars(
      chartData,
      groupScale,
      keys,
      labelSelector,
      [],
      range,
      xScale,
      yScale,
      'horizontal',
      colors
    );

    expect(result).toMatchSnapshot();
  });
});
