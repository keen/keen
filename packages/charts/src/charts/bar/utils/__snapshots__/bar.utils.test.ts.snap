// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculateGroupedBars() should calculate grouped bars correctly for horizontal layout 1`] = `
Array [
  Object {
    "color": "red",
    "colorOutOfRange": false,
    "height": 8.333333333333334,
    "key": "0.users",
    "selector": Array [
      0,
      "users",
    ],
    "value": 3,
    "width": 2.6785714285714306,
    "x": 26.785714285714285,
    "y": 75,
  },
  Object {
    "color": "green",
    "colorOutOfRange": false,
    "height": 8.333333333333334,
    "key": "0.licenses",
    "selector": Array [
      0,
      "licenses",
    ],
    "value": 52,
    "width": 46.42857142857142,
    "x": 26.785714285714285,
    "y": 83.33333333333333,
  },
  Object {
    "color": "blue",
    "colorOutOfRange": false,
    "height": 8.333333333333334,
    "key": "0.shops",
    "selector": Array [
      0,
      "shops",
    ],
    "value": 12,
    "width": 10.714285714285715,
    "x": 26.785714285714285,
    "y": 91.66666666666667,
  },
  Object {
    "color": "red",
    "colorOutOfRange": false,
    "height": 8.333333333333334,
    "key": "1.users",
    "selector": Array [
      1,
      "users",
    ],
    "value": 19,
    "width": 16.964285714285715,
    "x": 26.785714285714285,
    "y": 50,
  },
  Object {
    "color": "green",
    "colorOutOfRange": false,
    "height": 8.333333333333334,
    "key": "1.licenses",
    "selector": Array [
      1,
      "licenses",
    ],
    "value": 82,
    "width": 73.21428571428572,
    "x": 26.785714285714285,
    "y": 58.333333333333336,
  },
  Object {
    "color": "blue",
    "colorOutOfRange": false,
    "height": 8.333333333333334,
    "key": "1.shops",
    "selector": Array [
      1,
      "shops",
    ],
    "value": 15,
    "width": 13.392857142857146,
    "x": 26.785714285714285,
    "y": 66.66666666666667,
  },
  Object {
    "color": "red",
    "colorOutOfRange": false,
    "height": 8.333333333333334,
    "key": "2.users",
    "selector": Array [
      2,
      "users",
    ],
    "value": 20,
    "width": 17.85714285714286,
    "x": 26.785714285714285,
    "y": 25,
  },
  Object {
    "color": "green",
    "colorOutOfRange": false,
    "height": 8.333333333333334,
    "key": "2.licenses",
    "selector": Array [
      2,
      "licenses",
    ],
    "value": -15,
    "width": 13.392857142857142,
    "x": 13.392857142857142,
    "y": 33.333333333333336,
  },
  Object {
    "color": "blue",
    "colorOutOfRange": false,
    "height": 8.333333333333334,
    "key": "2.shops",
    "selector": Array [
      2,
      "shops",
    ],
    "value": 23,
    "width": 20.535714285714285,
    "x": 26.785714285714285,
    "y": 41.66666666666667,
  },
  Object {
    "color": "red",
    "colorOutOfRange": false,
    "height": 8.333333333333334,
    "key": "3.users",
    "selector": Array [
      3,
      "users",
    ],
    "value": 63,
    "width": 56.25000000000001,
    "x": 26.785714285714285,
    "y": 0,
  },
  Object {
    "color": "green",
    "colorOutOfRange": false,
    "height": 8.333333333333334,
    "key": "3.licenses",
    "selector": Array [
      3,
      "licenses",
    ],
    "value": -15,
    "width": 13.392857142857142,
    "x": 13.392857142857142,
    "y": 8.333333333333334,
  },
  Object {
    "color": "blue",
    "colorOutOfRange": false,
    "height": 8.333333333333334,
    "key": "3.shops",
    "selector": Array [
      3,
      "shops",
    ],
    "value": -30,
    "width": 26.785714285714285,
    "x": 0,
    "y": 16.666666666666668,
  },
]
`;

exports[`calculateGroupedBars() should calculate grouped bars correctly for vertical layout 1`] = `
Array [
  Object {
    "color": "red",
    "colorOutOfRange": false,
    "height": 2.6785714285714306,
    "key": "0.users",
    "selector": Array [
      0,
      "users",
    ],
    "value": 3,
    "width": 8.333333333333334,
    "x": 0,
    "y": 29.464285714285715,
  },
  Object {
    "color": "green",
    "colorOutOfRange": false,
    "height": 46.42857142857142,
    "key": "0.licenses",
    "selector": Array [
      0,
      "licenses",
    ],
    "value": 52,
    "width": 8.333333333333334,
    "x": 8.333333333333334,
    "y": 73.21428571428571,
  },
  Object {
    "color": "blue",
    "colorOutOfRange": false,
    "height": 10.714285714285715,
    "key": "0.shops",
    "selector": Array [
      0,
      "shops",
    ],
    "value": 12,
    "width": 8.333333333333334,
    "x": 16.666666666666668,
    "y": 37.5,
  },
  Object {
    "color": "red",
    "colorOutOfRange": false,
    "height": 16.964285714285715,
    "key": "1.users",
    "selector": Array [
      1,
      "users",
    ],
    "value": 19,
    "width": 8.333333333333334,
    "x": 25,
    "y": 43.75,
  },
  Object {
    "color": "green",
    "colorOutOfRange": false,
    "height": 73.21428571428572,
    "key": "1.licenses",
    "selector": Array [
      1,
      "licenses",
    ],
    "value": 82,
    "width": 8.333333333333334,
    "x": 33.333333333333336,
    "y": 100,
  },
  Object {
    "color": "blue",
    "colorOutOfRange": false,
    "height": 13.392857142857146,
    "key": "1.shops",
    "selector": Array [
      1,
      "shops",
    ],
    "value": 15,
    "width": 8.333333333333334,
    "x": 41.66666666666667,
    "y": 40.17857142857143,
  },
  Object {
    "color": "red",
    "colorOutOfRange": false,
    "height": 17.85714285714286,
    "key": "2.users",
    "selector": Array [
      2,
      "users",
    ],
    "value": 20,
    "width": 8.333333333333334,
    "x": 50,
    "y": 44.642857142857146,
  },
  Object {
    "color": "green",
    "colorOutOfRange": false,
    "height": 13.392857142857142,
    "key": "2.licenses",
    "selector": Array [
      2,
      "licenses",
    ],
    "value": -15,
    "width": 8.333333333333334,
    "x": 58.333333333333336,
    "y": 26.785714285714285,
  },
  Object {
    "color": "blue",
    "colorOutOfRange": false,
    "height": 20.535714285714285,
    "key": "2.shops",
    "selector": Array [
      2,
      "shops",
    ],
    "value": 23,
    "width": 8.333333333333334,
    "x": 66.66666666666667,
    "y": 47.32142857142857,
  },
  Object {
    "color": "red",
    "colorOutOfRange": false,
    "height": 56.25000000000001,
    "key": "3.users",
    "selector": Array [
      3,
      "users",
    ],
    "value": 63,
    "width": 8.333333333333334,
    "x": 75,
    "y": 83.03571428571429,
  },
  Object {
    "color": "green",
    "colorOutOfRange": false,
    "height": 13.392857142857142,
    "key": "3.licenses",
    "selector": Array [
      3,
      "licenses",
    ],
    "value": -15,
    "width": 8.333333333333334,
    "x": 83.33333333333333,
    "y": 26.785714285714285,
  },
  Object {
    "color": "blue",
    "colorOutOfRange": false,
    "height": 26.785714285714285,
    "key": "3.shops",
    "selector": Array [
      3,
      "shops",
    ],
    "value": -30,
    "width": 8.333333333333334,
    "x": 91.66666666666667,
    "y": 26.785714285714285,
  },
]
`;
